velox/docs/develop/tracing.rst

```diff
@@ -0,0 +1,152 @@
+
+1. **Query Trace Writer**: metadata writer and the data writer.
+2. **Query Trace Reader**: metadata reader and the data reader.
+3. **Query Trace Tool**: display query summaries or replay the
+   execution of the target operator.
+
+Query Trace Writer
+^^^^^^^^^^^^^^^^^^
+
+**QueryMetadataWriter** records the query metadata during task creation,
+serializes, and writes them into a file in JSON format. There are two kinds
+of metadata:
+
+- Query configurations and connector properties are specified by the user per query.
+  They can be serialized as JSON map objects (key-value pairs).
+- Plan fragment of the task (also known as a plan node tree). It can be serialized
+  as a JSON object, which is already supported in Velox.
+
+**QueryDataWriter** records the input vectors from the target operator, which are
+serialized and written as a binary file.
+
+Query Trace Reader
+^^^^^^^^^^^^^^^^^^
+
+**QueryMetadataReader** can load the query metadata JSON file, and extract the query
+configurations, connector properties, and the plan fragment.
+
+**QueryDataReader** can read and deserialize the input vectors of the target operator.
+It is used as the utility to replay the input data as a source operator in the target
+operator replay.
+
+**NOTE**: `QueryDataWriter` serializes and flushes the input vectors in batches,
+allowing it to replay the input process using the same sequence of batches.
+
+Query Trace Util
+^^^^^^^^^^^^^^^^
+
+- Create tracing directories.
+- Get query summaries.
+- Provide utilities to extract the target plan node, and assemble a plan fragment with
+  customized trace source and sink node with it.
+
+Query Trace Tool
+----------------
+
+The query trace tool leverages the trace reader to display query summaries and replay the
+execution of specific operators (TBD).
+
+Tracing tools Usage
+^^^^^^^^^^^^^^^^^^^
+
+.. code-block:: c++
+
+  query_trace_tool --root $root_dir --summary --pretty
+
+
+It would show something as the follows
+
+.. code-block:: c++
+
+  ++++++Query trace summary++++++
+  Number of tasks: 1
+  Task ids: task-1
+  ++++++Query configs and plan:++++++
+  {
+  "planNode":{
+  "filter":{},
+  "outputType":{...},
+  "nullAware":false,
+  "sources":[{...}, {...}],
+  "leftKeys":[],
+  "joinType":"INNER",
+  "id":"5",
+  "name":"HashJoinNode"
+  },
+  "connectorProperties":{...},
+  "queryConfig":{...}
+  }
+
+
+Here is a full list of supported command line arguments.
+
+* ``--usage``: Show the usage.
+* ``--root``: Root dir of the query tracing.
+* ``--summary``: Show the summary of the tracing including number of tasks and task ids.
+  It also print the query metadata including query configs, connectors properties, and query plan in JSON format.
+* ``--short_summary``: Only show number of tasks and task ids.
+* ``--pretty``: Show the summary of the tracing in pretty JSON.
+* ``--task_id``: Specify the target task id, if empty, show the summary of all the traced query tasks.
+
+
+Future Work
```

Comment:
Can we refer to GitHub issue here?

Reply:
That reference is good. Let's make sure to keep the issue updated with a list of completed and planned steps in the description. Thanks.